### DISCUSSION TITLE
Bug Or User Error - Not Sure - OTELHTTP transport not propagating Span and Trace IDs and leaving Parent blank

### DISCUSSION THREAD

**Discussion by js10x** [open-telemetry/opentelemetry-go#5896] [General]
### Description

OTELHTTP transport not propagating trace and span IDs, and is leaving the parent blank. The image shows the text details. 

![image](https://github.com/user-attachments/assets/f02f1b9b-c31c-490f-8711-883c9970d135)

### Environment

- OS: Windows 10
- Architecture: Any
- Go Version: 1.22
- opentelemetry-go version: latest

### Steps To Reproduce

1. Inject transport inside of http.Client
`http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}`

2. Use request with context
`req = req.WithContext(ctx)`

3. Fire off multiple requests with the same context and notice that new span and trace IDs are generated for each request.

### Expected behavior

I would expect to have the trace and span IDs propagate down from first HTTP request to all subsequent requests. The parent trace should also be filled out.

**Comment by pellared** [comment]
Can you please [create a Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces the issue you are encountering?

**Comment by js10x** [comment]
> Can you please [create a Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces the issue you are encountering?

I'll do ya one better, here's a full program to just copy paste right in a main.go file. All you have to do is run it after running go mod tidy and such. 
[main.txt](https://github.com/user-attachments/files/17403118/main.txt)

**Comment by dmathieu** [comment]
In the example you provided, the HTTP span has no parent, since no span is created before it. So the parent being empty/invalid is the expected behavior.

If I create a span before making the HTTP request:

```diff
diff --git a/tmp/main.go b/tmp/main.go
index 223283e..91668e8 100644
--- a/tmp/main.go
+++ b/tmp/main.go
@@ -25,8 +25,10 @@ func main() {
 func issueRequests() error {
        urls := []string{"https://www.google.com", "https://www.facebook.com", "https://www.yahoo.com"}
        client := &http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
+       tracer := otel.Tracer("main")

-       ctx := context.TODO()
+       ctx, span := tracer.Start(context.Background(), "issueRequests")
+       defer span.End()
        for _, v := range urls {
                req, err := http.NewRequestWithContext(ctx, http.MethodGet, v, nil)
                if err != nil {
```

Then, the span created by otelhttp has a parent, and that parent is indeed displayed by the exporter:

<details>
<summary>JSON Output</summary>

```json
﻿{
        "Name": "HTTP GET",
        "SpanContext": {
                "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
                "SpanID": "a3a139a7ac43a6bf",
                "TraceFlags": "01",
                "TraceState": "",
                "Remote": false
        },
        "Parent": {
                "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
                "SpanID": "8d341b5494fe4f9e",
                "TraceFlags": "01",
                "TraceState": "",
                "Remote": false
        },
        "SpanKind": 3,
        "StartTime": "2024-10-17T11:05:40.829184+02:00",
        "EndTime": "2024-10-17T11:05:40.946134625+02:00",
        "Attributes": [
                {
                        "Key": "http.method",
                        "Value": {
                                "Type": "STRING",
                                "Value": "GET"
                        }
                },
                {
                        "Key": "http.url",
                        "Value": {
                                "Type": "STRING",
                                "Value": "https://www.google.com"
                        }
                },
                {
                        "Key": "net.peer.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "www.google.com"
                        }
                },
                {
                        "Key": "http.status_code",
                        "Value": {
                                "Type": "INT64",
                                "Value": 200
                        }
                }
        ],
        "Events": null,
        "Links": null,
        "Status": {
                "Code": "Unset",
                "Description": ""
        },
        "DroppedAttributes": 0,
        "DroppedEvents": 0,
        "DroppedLinks": 0,
        "ChildSpanCount": 0,
        "Resource": [
                {
                        "Key": "service.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "unknown_service:main"
                        }
                },
                {
                        "Key": "telemetry.sdk.language",
                        "Value": {
                                "Type": "STRING",
                                "Value": "go"
                        }
                },
                {
                        "Key": "telemetry.sdk.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "opentelemetry"
                        }
                },
                {
                        "Key": "telemetry.sdk.version",
                        "Value": {
                                "Type": "STRING",
                                "Value": "1.31.0"
                        }
                }
        ],
        "InstrumentationScope": {
                "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
                "Version": "0.56.0",
                "SchemaURL": ""
        },
        "InstrumentationLibrary": {
                "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
                "Version": "0.56.0",
                "SchemaURL": ""
        }
}
{
        "Name": "HTTP GET",
        "SpanContext": {
                "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
                "SpanID": "92a4d6f4f87840e5",
                "TraceFlags": "01",
                "TraceState": "",
                "Remote": false
        },
        "Parent": {
                "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
                "SpanID": "8d341b5494fe4f9e",
                "TraceFlags": "01",
                "TraceState": "",
                "Remote": false
        },
        "SpanKind": 3,
        "StartTime": "2024-10-17T11:05:40.94616+02:00",
        "EndTime": "2024-10-17T11:05:41.230872334+02:00",
        "Attributes": [
                {
                        "Key": "http.method",
                        "Value": {
                                "Type": "STRING",
                                "Value": "GET"
                        }
                },
                {
                        "Key": "http.url",
                        "Value": {
                                "Type": "STRING",
                                "Value": "https://www.facebook.com"
                        }
                },
                {
                        "Key": "net.peer.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "www.facebook.com"
                        }
                },
                {
                        "Key": "http.status_code",
                        "Value": {
                                "Type": "INT64",
                                "Value": 200
                        }
                }
        ],
        "Events": null,
        "Links": null,
        "Status": {
                "Code": "Unset",
                "Description": ""
        },
        "DroppedAttributes": 0,
        "DroppedEvents": 0,
        "DroppedLinks": 0,
        "ChildSpanCount": 0,
        "Resource": [
                {
                        "Key": "service.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "unknown_service:main"
                        }
                },
                {
                        "Key": "telemetry.sdk.language",
                        "Value": {
                                "Type": "STRING",
                                "Value": "go"
                        }
                },
                {
                        "Key": "telemetry.sdk.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "opentelemetry"
                        }
                },
                {
                        "Key": "telemetry.sdk.version",
                        "Value": {
                                "Type": "STRING",
                                "Value": "1.31.0"
                        }
                }
        ],
        "InstrumentationScope": {
                "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
                "Version": "0.56.0",
                "SchemaURL": ""
        },
        "InstrumentationLibrary": {
                "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
                "Version": "0.56.0",
                "SchemaURL": ""
        }
}
```
</details>

**Comment by js10x** [comment]
Oh ok I see. None of the documentation samples pointed this out, so I was wondering if this was something I was missing. Appreciate it, thanks!

**Comment by js10x** [comment]
> In the example you provided, the HTTP span has no parent, since no span is created before it. So the parent being empty/invalid is the expected behavior.
> 
> If I create a span before making the HTTP request:
> 
> ```diff
> diff --git a/tmp/main.go b/tmp/main.go
> index 223283e..91668e8 100644
> --- a/tmp/main.go
> +++ b/tmp/main.go
> @@ -25,8 +25,10 @@ func main() {
>  func issueRequests() error {
>         urls := []string{"https://www.google.com", "https://www.facebook.com", "https://www.yahoo.com"}
>         client := &http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
> +       tracer := otel.Tracer("main")
> 
> -       ctx := context.TODO()
> +       ctx, span := tracer.Start(context.Background(), "issueRequests")
> +       defer span.End()
>         for _, v := range urls {
>                 req, err := http.NewRequestWithContext(ctx, http.MethodGet, v, nil)
>                 if err != nil {
> ```
> 
> Then, the span created by otelhttp has a parent, and that parent is indeed displayed by the exporter:
> 
> JSON Output
> ```json
> {
>         "Name": "HTTP GET",
>         "SpanContext": {
>                 "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
>                 "SpanID": "a3a139a7ac43a6bf",
>                 "TraceFlags": "01",
>                 "TraceState": "",
>                 "Remote": false
>         },
>         "Parent": {
>                 "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
>                 "SpanID": "8d341b5494fe4f9e",
>                 "TraceFlags": "01",
>                 "TraceState": "",
>                 "Remote": false
>         },
>         "SpanKind": 3,
>         "StartTime": "2024-10-17T11:05:40.829184+02:00",
>         "EndTime": "2024-10-17T11:05:40.946134625+02:00",
>         "Attributes": [
>                 {
>                         "Key": "http.method",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "GET"
>                         }
>                 },
>                 {
>                         "Key": "http.url",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "https://www.google.com"
>                         }
>                 },
>                 {
>                         "Key": "net.peer.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "www.google.com"
>                         }
>                 },
>                 {
>                         "Key": "http.status_code",
>                         "Value": {
>                                 "Type": "INT64",
>                                 "Value": 200
>                         }
>                 }
>         ],
>         "Events": null,
>         "Links": null,
>         "Status": {
>                 "Code": "Unset",
>                 "Description": ""
>         },
>         "DroppedAttributes": 0,
>         "DroppedEvents": 0,
>         "DroppedLinks": 0,
>         "ChildSpanCount": 0,
>         "Resource": [
>                 {
>                         "Key": "service.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "unknown_service:main"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.language",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "go"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "opentelemetry"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.version",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "1.31.0"
>                         }
>                 }
>         ],
>         "InstrumentationScope": {
>                 "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
>                 "Version": "0.56.0",
>                 "SchemaURL": ""
>         },
>         "InstrumentationLibrary": {
>                 "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
>                 "Version": "0.56.0",
>                 "SchemaURL": ""
>         }
> }
> {
>         "Name": "HTTP GET",
>         "SpanContext": {
>                 "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
>                 "SpanID": "92a4d6f4f87840e5",
>                 "TraceFlags": "01",
>                 "TraceState": "",
>                 "Remote": false
>         },
>         "Parent": {
>                 "TraceID": "f36f3a4ea90e9b994c19a3cd9dbbdfbe",
>                 "SpanID": "8d341b5494fe4f9e",
>                 "TraceFlags": "01",
>                 "TraceState": "",
>                 "Remote": false
>         },
>         "SpanKind": 3,
>         "StartTime": "2024-10-17T11:05:40.94616+02:00",
>         "EndTime": "2024-10-17T11:05:41.230872334+02:00",
>         "Attributes": [
>                 {
>                         "Key": "http.method",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "GET"
>                         }
>                 },
>                 {
>                         "Key": "http.url",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "https://www.facebook.com"
>                         }
>                 },
>                 {
>                         "Key": "net.peer.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "www.facebook.com"
>                         }
>                 },
>                 {
>                         "Key": "http.status_code",
>                         "Value": {
>                                 "Type": "INT64",
>                                 "Value": 200
>                         }
>                 }
>         ],
>         "Events": null,
>         "Links": null,
>         "Status": {
>                 "Code": "Unset",
>                 "Description": ""
>         },
>         "DroppedAttributes": 0,
>         "DroppedEvents": 0,
>         "DroppedLinks": 0,
>         "ChildSpanCount": 0,
>         "Resource": [
>                 {
>                         "Key": "service.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "unknown_service:main"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.language",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "go"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.name",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "opentelemetry"
>                         }
>                 },
>                 {
>                         "Key": "telemetry.sdk.version",
>                         "Value": {
>                                 "Type": "STRING",
>                                 "Value": "1.31.0"
>                         }
>                 }
>         ],
>         "InstrumentationScope": {
>                 "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
>                 "Version": "0.56.0",
>                 "SchemaURL": ""
>         },
>         "InstrumentationLibrary": {
>                 "Name": "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",
>                 "Version": "0.56.0",
>                 "SchemaURL": ""
>         }
> }
> ```

Is there any way to suppress logging for just the root span? I only want the children of the root span to be logged

**Comment by dmathieu** [comment]
You would have to setup a custom exporter, or processor to skip some spans from being exported.
The stdout exporter is only meant for debugging, not for production usage. So this kind of need isn't really planned for that exporter.

**Comment by js10x** [comment]
Stderr is supposed to be used yes? Also, I have a dumb question: I've created and injected a custom log exporter by implementing the /sdk/log -> Exporter interface, and passing in the new impl to the logger provider, but I can't seem to figure out how to catch emitted logs. Are they emitted every time a consuming application writes log.Println(), or fmt.Printn()?

**Comment by dmathieu** [comment]
> Stderr is supposed to be used yes?

Not in production. Only for light debugging.

> Are they emitted every time a consuming application writes log.Println(), or fmt.Printn()?

I invite you to look at the doc and implementation.
If you use [BatchProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/log#NewBatchProcessor), your logs will be exported at regular intervals.
If you use [SimpleProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/log#SimpleProcessor), they will be exporter synchronously as soon as they are emitted.

You may want to have a look at the [dice example](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/examples/dice), which is a sample app that emits every signal.

**Reply by js10x** [reply]
> > Stderr is supposed to be used yes?
> 
> Not in production. Only for light debugging.
>
Oh really, now I'm confused haha. My goal is to capture all writes to the console and append custom attributes to the logs, which is what I thought the exporter did, but maybe I'm wrong? So, if stdout and stderr is not used to show output in PROD, where are the logs actually exported to? 
 
> > Are they emitted every time a consuming application writes log.Println(), or fmt.Printn()?
> 
> I invite you to look at the doc and implementation. If you use [BatchProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/log#NewBatchProcessor), your logs will be exported at regular intervals. If you use [SimpleProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/log#SimpleProcessor), they will be exporter synchronously as soon as they are emitted.
> 
> You may want to have a look at the [dice example](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/examples/dice), which is a sample app that emits every signal.

I did read through those examples in getting started. I've injected my custom exporter into the provider. So then when there is a trace, I'm not seeing that the Export method is called at all. Am I missing something? 

![image](https://github.com/user-attachments/assets/07f27a03-3581-48dd-876f-0a9313e08602)

**Comment by js10x** [comment]
Figured this out. In case someone else should happen upon this thread, I'll explain what I was missing. The "Export" method wasn't being called b/c I didn't know that I had to use a logger (slog.Logger is what I used) that is instrumented with an OTEL handler. Once configured all calls to slogger.Log(..) will trigger the "Export" method (depending on your processor, simple will call export each time a record is created). I was thinking that all calls to the log.Println(..) would trigger this method, but it does not.
